File: Задачи (по источникам)/Summer School (NIS, 2024)/Day 2 + Day 3/B/B.cpp
// author: Danila "akshin_" Axyonov

#include <iostream>
using namespace std;
typedef long long int lli;

#define plli                pair<lli, lli>
#define mp(_first, _second) make_pair(_first, _second)
#define pb(_elem)           push_back(_elem)

const lli INF = (lli)1e18 + 1LL;

lli n;
lli dl = INF,
    dr = INF;
lli m;
lli ans;

void solve();
lli sum_of_digits(lli);
bool is_prime(lli);

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    solve();
    
    return 0;
}

void solve() {
    cin >> n;
    
    m = n;
    while (m > 0LL) {
        if (is_prime(sum_of_digits(m))) {
            dl = abs(n - m);
            break;
        }
        --m;
    }

    m = n;
    while (true) {
        if (is_prime(sum_of_digits(m))) {
            dr = abs(n - m);
            break;
        }
        ++m;
    }

    ans = min(dl, dr);
    cout << ans << '\n';
    return;
}

lli sum_of_digits(lli num) {
    lli res = 0LL;
    while (num) {
        res += num % 10LL;
        num /= 10LL;
    }
    return res;
}

bool is_prime(lli num) {
    if (num == 1LL)
        return false;
    for (lli d = 2LL; d * d <= num; ++d)
        if (num % d == 0LL)
            return false;
    return true;
}

File: Задачи (по источникам)/Summer School (NIS, 2024)/Day 2 + Day 3/B/B.py
def is_prime(num):
    if num == 1:
        return False
    d = 2
    while d ** 2 <= num:
        if num % d == 0:
            return False
        d += 1
    return True

n = int(input())

dl = 10 ** 20
m = n
while m > 0:
    digits = list(map(int, str(m)))
    digits_sum = sum(digits)
    if is_prime(digits_sum):
        dl = abs(n - m)
        break
    m -= 1

dr = 10 ** 20
m = n
while True:
    digits = list(map(int, str(m)))
    digits_sum = sum(digits)
    if is_prime(digits_sum):
        dr = abs(n - m)
        break
    m += 1

ans = min(dl, dr)
print(ans)
